src/inmanta/deploy/scheduler.py

```diff
@@ -532,6 +545,9 @@ async def _new_version(
                 for resource in blocked_resources | transitively_blocked_resources:
                     self._work.delete_resource(resource)
 
+                # Write dirty state back to the database
+                await ResourcePersistentState.mark_as_has_update(self.environment, set(self._state.dirty))
```

Comment:
I'm hesitant about doing this under the scheduler lock, because it is a potentially slow IO operation, and everything under the lock is quite performance sensitive.

But doing it outside the lock also requires a lot of care. I wonder if we should consider bulk db writes every few seconds instead of on-the-fly writes. I'll come back to this.

Reply:
This feels like a premature optimization. Like you say: _"it is a **potentially** slow IO operation"_. Wouldn't it be better to keep it like this an evaluate the performance. We can still optimize this later on, right? Or do you see issues with that?

Reply:
What I mean with "potentially" in this case is not "we don't know if it will be", but more "it occasionally *is*", e.g. waiting for a connection or a db lock. I'll try to give this some more thought today, but at the moment I think we should try to find a way where our db writes do not block scheduler operation, e.g. we read everything we want to flush inside the lock, but do the actual flushing outside of it. We then only have to make sure that our flushes can never be done out of order.